### PR TITLE
Introducing foundry tests

### PR DESCRIPTION
**Pull request by 0x-r4bbit**
This PR introduces foundry to the smart contracts codebase.
It does **not** replace the hardhat setup because

1. this is meant to be an experiment and there should be a chance for developers to try things out
2. deployment scripts have not yet ported over to foundry

That being said, I'm aiming to keep test parity between the hardhat and foundry setup.
This PR also introduces a CI setup so that tests are executed on each PR.

All in all foundry tests run much faster. Also, writing tests in foundry seems much more pleasant as tests can be written in Solidity. So there's no need for things like bigint conversions.

In addition, foundry offers a good battery of EVM tools to make building and testing easier.

### CODE REVIEW

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -21,3 +25,31 @@ jobs:
...
+          submodules: recursive
+
+      - name: Install Foundry
+        uses: foundry-rs/foundry-toolchain@v1
+        with:
+          version: nightly
+
+      - name: Run Forge build
+        run: |
+          cd packages/contracts
```
Comment (0x-r4bbit) [inline .github/workflows/CI.yml:46]: There's probably a better way to navigate here once and then preserve CWD across tasks, but it does the trick for now.

File: packages/contracts/.github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      - name: Run Forge build
+        run: |
+          forge --version
+          forge build --sizes
+        id: build
+
+      - name: Run Forge tests
+        run: |
+          forge test -vvv
+        id: test
```
Comment (0x-r4bbit) [inline packages/contracts/.github/workflows/test.yml:34]: This one came with `forge init` and can be ignored for now. I've moved the relevant bits to the top level workflow file.

File: packages/contracts/test/FeaturedVotingContract.t.sol [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+        votes[0] = _createSignedVote(bobsKey, alice, communityID1, 200, block.timestamp);
+
+        vm.expectEmit(false, false, false, true);
+        emit InvalidSignature(communityID1, alice);
+        featuredVotingContract.castVotes(votes);
+    }
+
+    // // it seems impossible to come up with a scenario in which `castVotes()`
+    // // would emit `CommunityFeaturedRecently`, so this test is disabled
+    // function test_CastVotes_CooldownPeriodHasNotPassed() public {
```
Comment (0x-r4bbit) [inline packages/contracts/test/FeaturedVotingContract.t.sol:314]: @osmaczko would be wrong but it look like we can remove some code in `castVotes()`.
I couldn't get a scenario to work in which `CommunityFeaturedRecently` would be emitted via `castVotes`.

**Review by 0x-r4bbit — COMMENTED** [review]
@osmaczko I've been updating this branch and also added CI integration.
This PR does not disable any of the hardhat stuff (yet) so there's a chance for a transitioning phase.

Still need to look into deployment scripts with foundry, so until that is figured out I'd rather have this at most side by side.

File: packages/contracts/test/FeaturedVotingContract.t.sol (hunk 3 above)
Comment (osmaczko) [reply packages/contracts/test/FeaturedVotingContract.t.sol:358]: https://github.com/status-im/community-dapp/pull/60

File: packages/contracts/.github/workflows/test.yml
[the remark this answers is not in the document]
Comment (0x-r4bbit) [reply packages/contracts/.github/workflows/test.yml:1]: Yea as mentioned in my comment, it's the CI file you get out of the box when you generate a foundry project. I left it here for completeness-sake but can remove it as I've also copied the important bits into the toplevel CI file.

File: packages/contracts/lib/forge-std
[the remark this answers is not in the document]
Comment (0x-r4bbit) [reply packages/contracts/lib/forge-std:1]: Unfortunately, that's what foundry does at the moment. Can most likely use `npm` for installing from github paths, but then we need to ensure we always install them in the right places (in this case it's the `lib` folder that foundry expects.

So it's either a) go with whatever foundry uses atm and don't worry about configuration or b) do your own thing but then requires a bit of custom setup

File: packages/contracts/test/FeaturedVotingContract.t.sol (hunk 3 above)
Comment (0x-r4bbit) [reply packages/contracts/test/FeaturedVotingContract.t.sol:358]: Cool, will review and update this PR accordingly if it's the right path forward

File: packages/contracts/lib/forge-std
[the remark this answers is not in the document]
Comment (osmaczko) [reply packages/contracts/lib/forge-std:1]: That's a pity :sweat_smile: The former (a) sounds much more reasonable.

File: packages/contracts/.github/workflows/test.yml
[the remark this answers is not in the document]
Comment (osmaczko) [reply packages/contracts/.github/workflows/test.yml:1]: I would opt for removing it, it is not used, right?

File: packages/contracts/test/FeaturedVotingContract.t.sol (hunk 3 above)
Comment (0x-r4bbit) [reply packages/contracts/test/FeaturedVotingContract.t.sol]: Tested your PR and ran the enabled foundry test on top of it as well.
Everything's green!

File: packages/contracts/.github/workflows/test.yml
[the remark this answers is not in the document]
Comment (0x-r4bbit) [reply packages/contracts/.github/workflows/test.yml:1]: Removed the unused CI file

**Review by osmaczko — APPROVED** [review]

**Pull request merged**